### PR TITLE
gh-118090: Improve error message for empty type param brackets

### PR DESCRIPTION
**Pull request by sobolevn**
* Issue: gh-118090

### CODE REVIEW

File: Grammar/python.gram
Comment (AlexWaygood) [inline Grammar/python.gram]: Should this maybe be "type parameter definition" rather than "type variable definition"? TypeVarTuples and ParamSpecs can also be defined within the square brackets

**Comment by sobolevn** [comment]
Ubuntu failures are unrelated: connection timeout.

File: Grammar/python.gram
Comment (JelleZijlstra) [inline Grammar/python.gram]: I think this is clearer and more succinct:
```suggestion
            "Type parameter list may not be empty")}
```

**Review by JelleZijlstra — COMMENTED** [review]
Thanks! Left a suggestion for improving the error.

File: Grammar/python.gram
Comment (AlexWaygood) [reply Grammar/python.gram]: "May not" sounds a bit "weak" to me... how about:

```suggestion
            "Type parameter list cannot be empty")}
```

**Review by JelleZijlstra — APPROVED** [review]

**Comment by sobolevn** [comment]
Thanks for the review! I will wait for @pablogsal or @lysnikolaou review as code owners :)

File: Grammar/python.gram
Comment (pablogsal) [inline Grammar/python.gram]: One small nit: can you please move this to an invalid rule? There are some optimisations so this is not checked during the first pass

Comment (pablogsal) [reply Grammar/python.gram]: Check https://devguide.python.org/internals/parser/#grammatical-elements-and-rules for more info on this

Comment (sobolevn) [reply Grammar/python.gram]: Interesting, when I move to the individual rule, this happens:

```diff
diff --git Grammar/python.gram Grammar/python.gram
index c29e9450499..14f2a847c97 100644
--- Grammar/python.gram
+++ Grammar/python.gram
@@ -639,12 +639,9 @@ type_alias[stmt_ty]:
 # --------------------------
 
 type_params[asdl_type_param_seq*]: 
+    | invalid_type_params
     | '[' t=type_param_seq  ']' {
         CHECK_VERSION(asdl_type_param_seq *, 12, "Type parameter lists are", t) }
-    | '[' token=']' {
-        RAISE_SYNTAX_ERROR_STARTING_FROM(
-            token, 
-            "Type parameter list cannot be empty")}
 
 type_param_seq[asdl_type_param_seq*]: a[asdl_type_param_seq*]=','.type_param+ [','] { a }
 
@@ -1429,3 +1426,9 @@ invalid_arithmetic:
     | sum ('+'|'-'|'*'|'/'|'%'|'//'|'@') a='not' b=inversion { RAISE_SYNTAX_ERROR_KNOWN_RANGE(a, b, "'not' after an operator must be parenthesized") }
 invalid_factor:
     | ('+' | '-' | '~') a='not' b=factor { RAISE_SYNTAX_ERROR_KNOWN_RANGE(a, b, "'not' after an operator must be parenthesized") }
+
+invalid_type_params:
+  | '[' token=']' {
+        RAISE_SYNTAX_ERROR_STARTING_FROM(
+            token, 
+            "Type parameter list cannot be empty")}
diff --git Parser/parser.c Parser/parser.c
index d64d08d1024..6cae7697423 100644
Binary files Parser/parser.c and Parser/parser.c differ

```

Test failure:
```
File "/Users/sobolev/Desktop/cpython2/Lib/test/test_syntax.py", line 2028, in test.test_syntax
Failed example:
    def some[](): ...
Expected:
    Traceback (most recent call last):
    ...
    SyntaxError: Type parameter list cannot be empty
Got:
    Traceback (most recent call last):
      File "/Users/sobolev/Desktop/cpython2/Lib/doctest.py", line 1385, in __run
        exec(compile(example.source, filename, "single",
             ~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                     compileflags, True), test.globs)
                     ^^^^^^^^^^^^^^^^^^^
      File "<doctest test.test_syntax[375]>", line 1
        def some[](): ...
                ^
    SyntaxError: expected '('


0:00:00 load avg: 2.18 [2/2/1] test_grammar -- test_syntax failed (1 failure)

== Tests result: FAILURE ==

1 test failed:
    test_syntax
```

Comment (pablogsal) [reply Grammar/python.gram]: This is because when you are moving the rule to the second phase there is another one with higher precedence that triggers the failure.

You need to find the common branch and move it so this is inspected before

Comment (sobolevn) [reply Grammar/python.gram]: This happens because of `&&'('` in `'def' n=NAME t=[type_params] &&'(' params=[params] ')' a=['->' z=expression { z }] &&':' tc=[func_type_comment] b=block`

Looks like that this forced token does not want to work with `invalid_` rules :thinking:
Finding some other ways.

Comment (sobolevn) [reply Grammar/python.gram]: @pablogsal would it make sense to move 

```python
>>> def f: 
  File "<stdin>", line 1
    def f: 
         ^
SyntaxError: expected '('
```

case to `invalid_def_raw`? And to remove `&&` from `'('`?

Comment (pablogsal) [reply Grammar/python.gram]: It makes sense to me as that would be less restricting

Comment (pablogsal) [reply Grammar/python.gram]: > Looks like that this forced token does not want to work with `invalid_` rules 🤔

That's because the forced token it's inspected on the first pass and takes precedence (it fails as soon as it is not satisfied)

Comment (sobolevn) [reply Grammar/python.gram]: Done!

**Review by lysnikolaou — APPROVED** [review]
LGTM! Makes sense we had to remove the forced token stuff from the first pass.

**Pull request merged**
